File: questions/110.BalancedBinaryTree.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

/**
 * isBalanced - DFS
 * @param {TreeNode} root
 * @return {boolean}
 */
const isBalanced = function (root) {
  /**
   * @param {TreeNode} node
   * @return {{isBalanced: boolean, height: number}}
   */
  const traverse = (node) => {
    if (!node) return { isBalanced: true, height: 0 }

    const left = traverse(node.left)
    const right = traverse(node.right)
    const isBalanced =
      left.isBalanced &&
      right.isBalanced &&
      Math.abs(left.height - right.height) <= 1

    return {
      isBalanced,
      height: 1 + Math.max(left.height, right.height),
    }
  }

  return traverse(root).isBalanced
}

const tree = new TreeNode(3)
tree.left = new TreeNode(9)
tree.right = new TreeNode(20)
tree.right.left = new TreeNode(15)
tree.right.right = new TreeNode(7)

console.log(isBalanced(tree))

const tree2 = new TreeNode(1)
tree2.left = new TreeNode(2)
tree2.right = new TreeNode(2)
tree2.left.left = new TreeNode(3)
tree2.left.right = new TreeNode(3)
tree2.left.left.left = new TreeNode(4)
tree2.left.left.right = new TreeNode(4)

console.log(isBalanced(tree2))

File: questions/110.BalancedBinaryTree.go
package questions

import "math"

type Treenode110 struct {
	Val   int
	Left  *Treenode110
	Right *Treenode110
}

type Traverse110 struct {
	Height     int
	IsBalanced bool
}

func isBalanced(root *Treenode110) bool {
	return traverse(root).IsBalanced
}
func traverse(node *Treenode110) Traverse110 {
	if node == nil {
		return Traverse110{Height: 0, IsBalanced: true}
	}

	left := traverse(node.Left)
	right := traverse(node.Right)
	b := left.IsBalanced && right.IsBalanced && math.Abs(float64(left.Height-right.Height)) <= 1

	return Traverse110{
		IsBalanced: b,
		Height:     1 + int(math.Max(float64(left.Height), float64(right.Height))),
	}
}

//tree := &TreeNode_110{Val: 3}
//tree.Left = &TreeNode_110{Val: 9}
//
//tree.Right = &TreeNode_110{Val: 20}
//tree.Right.Left = &TreeNode_110{Val: 15}
//tree.Right.Right = &TreeNode_110{Val: 7}
//
//result := isBalanced(tree)
//
//fmt.Printf("%+v", result)
